File: snakesolver.pl
#!perl6

use v6;

sub dont(&) {}

my $dim=4;
my @snake_short=<5 1 1 2 1 1 1 6 1 1 2 2 4 2 3 4 3 5 5 2 2 4 2 2 2>;
# my @snake_short=<7 1 1 2 1 1 1 6 1 1 2 2 4 2 3 4 3 5 5 2 2 4 2 2>;
# my @snake_short=<2 4 2 2 4 4 4 2 2 4 1 1 4 2 2 4 4 4 2 2 4 3 1>;

my $dimm1= $dim - 1;
my $dimp1= $dim + 1;
my $dimp2= $dim + 2;
my $dimp2sq= $dimp2 * $dimp2;

my $slen= $dim * $dim * $dim;

my $color= 2;
my @snake= @snake_short.map({ my $c= $color= 3 - $color; (^$_).map({ $c }); });

die "snake length is not $slen!" if $slen != @snake.elems;
die "snake length is uneven!" if @snake.elems ~& 1;


sub xyz2n($x, $y, $z) { $z * $dimp2sq + $y * $dimp2 + $x };
sub n2x($n) {  $n               % $dimp2 };
sub n2y($n) { ($n div $dimp2)   % $dimp2 };
sub n2z($n) { ($n div $dimp2sq) % $dimp2 };

sub getCube( @result ) {
    (1..$dim).map(sub ($z) {
        # z
        "\nz=$z\n" ~ (^$dimp1).map(sub ($y) {
            # y
            if $y == 0 || $y == $dimp1 {
                return "\t+----+----+----+----+\t+---+---+---+---+\n";
            }
            else {
                return "\t|" ~ (1..$dim).map(sub ($x) {
                    # x
                    my $n= xyz2n($x, $y, $z);
                    sprintf " %2d |", @result[$n];
                }).join('') ~ "\t|" ~ (1..$dim).map(sub ($x) {
                    # x
                    my $n= xyz2n($x, $y, $z);
                    if @snake[@result[$n]] == 2 {
                        " X |";
                    }
                    else {
                        "   |";
                    }
                }).join('') ~ "\n\t+----+----+----+----+\t+---+---+---+---+\n";
            }
        });
    });
};



my @result;
my @expected;

# prepare an empty cube with borders (@result) and define color of visible fronts (@expected)
for ^$dimp2 -> $x {
    for ^$dimp2 -> $y {
        for ^$dimp2 -> $z {
            my $n= xyz2n($x, $y, $z);
            if ($x, $y, $z).grep({ $_ == 0 || $_ == $dimp1 }) {
                # we are in outer space
                @result[$n]= -1;
                next;
            }
            if ($x, $y, $z).grep({ $_ == 1 || $_ == $dim }) {
                # we are at the surface
                @expected[$n]= ([+] ($x, $y, $z).map({ ($_ + 3) div 2 })) % 2 + 1;
                next;
            }
        }
    }
}

my $try= 0;

sub solve($n, $idx, @currentState is copy) {

    # return if new position is occupated already
    return if @currentState[$n];

    # return if special color is required but does not match
    return if @expected[$n] && @expected[$n] != @snake[$idx];

    @currentState[$n]= $idx;
    $try++;
    print "\r$try (" ~ n2x($n) ~ "," ~ n2y($n) ~ "," ~ n2z($n) ~ ") -> $idx: " ~ @snake[$idx] unless $try % 300;

    # we have solved it that way
    solved(@currentState) if $idx === @snake.elems;

    return (1, -1, $dimp2, -$dimp2, $dimp2sq, -$dimp2sq).map({ $n + $_ }).grep({ !@currentState[$_] }).map({ [$_, $idx + 1, [@currentState]] });
}

sub solved( @result ) {
    say "SOLVED! in $try tries";
    say getCube(@result);
}

for ^$slen {
    say;
    say "Starting over: $_";


    my @queue= ([xyz2n(1, 1, 1), 0, [@result]], );
    while @queue {
        @queue.unshift(solve(|@queue.shift));
    }

    @snake.push(@snake.shift);
}

File: snakesolver.py
#!/usr/bin/python

dim= 4
snake_short= [(5, 1), (1, 2), (1, 1), (1, 6), (1, 1), (2, 2), (4, 2), (3, 4), (3, 5), (5, 2), (2, 4), (2, 2), (2, 0)]
exit_after_first_solution= True

dimp1= dim + 1
dimp2= dim + 2
dimp2sq= dimp2 * dimp2
slen= dim * dim * dim

add= lambda a, b: a + b

snake= reduce(
    add,
    [
        [ 1 for x in range( s[0] ) ] + [ 2 for x in range( s[1] ) ] for s in snake_short
    ]
)

result= [
    [
        [
            -1 if len(filter(lambda _: _ == 0 or _ == dimp1, [x, y, z])) > 0 else 0 for z in range(dimp2) 
        ] for y in range(dimp2) 
    ] for x in range(dimp2)
]

expected= [
    [
        [
            reduce(
                add,
                map(lambda a: (a + 3) // 2, [x, y, z])
            ) % 2 + 1 if len(filter(lambda _: _ == 1 or _ == dim, [x, y, z])) > 0 else 0 for z in range(dimp2) 
        ] for y in range(dimp2)
    ] for x in range(dimp2)
]

if len(snake) != slen:
    print "Snake has length", len(snake), "but should have length", slen, "."
    sys.exit


tries= 0

def solveSnake(snake):
    def getCube(result):

        def getXresult(x, y, z):
            return ' {:2d} |'.format(result[x][y][z])

        def getXindex(x, y, z):
            return ' ' + ('X' if snake[result[x][y][z]] == 1 else ' ') + ' |'

        def getY(y, z):
            return '\t|' + reduce(
                add, [getXresult(x, y, z) for x in range(1, dimp1)]
            ) + '\t|' + reduce(
                add, [getXindex(x, y, z) for x in range(1, dimp1)]
            ) + '\n\t+' + reduce(
                add, ['----+' for x in range(1, dimp1)]
            ) + '\t+' + reduce(
                add, ['---+' for x in range(1, dimp1)]
            ) +'\n'

        def getZ(z):
            return '\nz= ' + str(z) + '\n\t+' + reduce(
                add, ['----+' for x in range(1, dimp1)]
            ) + '\t+' + reduce(
                add, ['---+' for x in range(1, dimp1)]
            ) + '\n' + reduce(
                add, [getY(y, z) for y in range(1, dimp1)]
            ) + '\n'

        return reduce(add, [getZ(z) for z in range(1, dimp1)])


    def solve(xyz, idx, currentState):

        if idx == slen:
            solved(currentState)
            return
#        print xyz, idx, snake[idx]

        getValueAt= lambda l, xyz: l[xyz[0]][xyz[1]][xyz[2]]
        getCurrentState= lambda xyz: getValueAt(currentState, xyz)
        addXyz= lambda add: [ _[0] + _[1] for _ in zip(xyz, add)]

        if getCurrentState(xyz) != 0:
#            print "->", getCurrentState(xyz)
            return

        if not getValueAt(expected, xyz) in set((0, snake[idx])):
#            print "=>", getValueAt(expected, xyz), snake[idx]
            return

        currentState[xyz[0]][xyz[1]][xyz[2]]= idx
        global tries
        tries+= 1
        
        if tries % 30000 == 0:
            print tries, xyz, " ->" , idx, ": ", snake[idx]
    
        for _xyz in [addXyz(_add) for _add in ((1, 0, 0), (-1, 0, 0), (0, 1, 0), (0, -1, 0), (0, 0, 1), (0, 0, -1))]:
            solve(_xyz, idx + 1, currentState)
    
        currentState[xyz[0]][xyz[1]][xyz[2]]= 0
    
    def solved(result):
        print "SOLVED", tries
        print getCube(result)
        if exit_after_first_solution:
            sys.exit
    
    solve((1, 1, 1), 0, result)



for i in range(slen):
    print "Starting over", i

    solveSnake(snake[i:] + snake[:i])
